source/s3_paginator.c

```diff
@@ -0,0 +1,453 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/s3/private/s3_paginator.h>
```

Comment:
Are this revamping the list_object to be shared between different operation? How much new code here?

Reply:
yeah, this is refactoring of old list_object code to break paginator out of it and reusing paginator for both list objects and parts. A lot of code is reused and its mostly just rewiring old code